Automattic/pocket-casts-android
# Warn when changing to metered network if setting is enabled

## Description

For users who have enabled the "Warn before using data" setting, we would check if a metered connection was being used when an episode started playing, but we would not do anything if, during playback, the user switched from an unmetered connection to a metered connection. This PR addresses that by detecting when the app has switched to a metered connection, and then pausing playback and presenting the user with a notification asking if they would like to continue playback.

We are only presenting this warning once when an episode is played. This means that if a user gets a warning when they start playing an episode, and they choose to stream it, then we will not present any later warnings if their network switches back and forth between metered networks. Likewise, if a user switches between metered and unmetered networks multiple times, we will only display the warning once per episode.

Fixes #1374
h/t to @neeldoshii for helping inspire this solution with https://github.com/Automattic/pocket-casts-android/pull/1447

These changes should not affect the watch or automotive apps because the warn before streaming setting defaults to false and cannot be changed within those apps. This is something we'll need to consider when we start syncing settings though.

## Testing Instructions

For all of these tests:
1. Install a build from this PR
3. Make sure you have two networks available: one that is treated as unmetered and one that is treated as metered.

### 1. Warns one time per episode when switching to metered network
2. Turn the setting for "Warn before using data" under "Storage & data use" to be on
1. Connect to an unmetered network
2. Begin streaming an episode
3. ✅  Verify that you do NOT get a stream warning
4. Switch to a metered network
5. ✅  Verify that you do get a stream warning and playback pauses
6. Tap the notification button to continue streaming ("Yes, keep playing")
7. ✅ Verify that playback continues
8. Switch to your unmetered network
9. Switch back to your metered network
10. ✅ Verify that you are NOT presented with a second stream warning
11. Begin streaming a different episde
12. ✅ Verify that you are shown a stream warning (whether you get a system notification or a bottom sheet depends on where you start playback from within the app and is not affected by the changes in this PR)
13. Tap the button to continue playing the episode
5. Switch to an unmetered network
6. Switch back to your metered network
7. ✅ Verify that you are NOT presented with a stream warning

### 2. Play next downloaded button works as before
2. Turn the setting for "Warn before using data" under "Storage & data use" to be on
1. Add a few episodes that are not downloaded to your queue
1. Add a downloaded episode to the middle of your queue
2. Take note of the top downloaded episode in your queue (it should not be at the top of your queue though)
1. Connect to an unmetered network
1. Begin streaming an episode
1. Switch to a metered network
1. ✅ Verify that you are shown a stream warning and playback pauses
1. Tap the "Play next downloaded" notification button
3. Verify that the episode that was previously the highest downloaded episode in your queue has started playing.

### 3. Does not show stream warning for downloaded episodes
2. Turn the setting for "Warn before using data" under "Storage & data use" to be on
2. Connect to an unmetered network
3. Download an episode
5. Begin playing that episode
6. Switch to a metered connection
8. ✅ Verify that you are NOT shown a stream warning

### 4. Does not show stream warning if setting is turned off
1. Turn the setting for "Warn before using data" under "Storage & data use" to be OFF
2. Connect to an unmetered network
3. Begin streaming an episode
4. Switch to an unmetered network
5. ✅ Verify that you are NOT shown a stream warning
6. Start streaming a different episode
9. ✅ Verify that playback starts immediately and you are NOT shown a stream warning.

### 5 Does not show stream warning if chromecasting
2. Turn the setting for "Warn before using data" under "Storage & data use" to be on
2. Connect to an unmetered network
3. Begin chromecasting an episode that you have not downloaded
4. Switch your device to a metered connection
5. ✅ Verify that you are NOT shown a stream warning


### 6. Fixes double-warning issue
2. Turn the setting for "Warn before using data" under "Storage & data use" to be on
3. Connect to a metered network
4. Make sure you have some episodes in your up next queue
5. Start playback from your up next queue by long-pressing on one of the items (or if you have the setting turned on for a regular press to start playback instead of opening the episode details screen, do that)
6. Observe that you get a notification asking if you want to stream
7. Proceed to stream the episode
8. Pause the episode from the miniplayer
9. Resume the episode from the miniplayer
10. ✅ Observe that the episode plays and you are not again presented with a notification asking whether you want to stream the episode (before this PR, you would get asked a second time in this flow).

## Screenshots or Screencast 

https://github.com/Automattic/pocket-casts-android/assets/4656348/c355bafb-790e-4ac4-b34a-1be78bf931be

## Checklist
